src/postprocess/chi0.jl

```diff
@@ -134,34 +132,138 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
     δψnk
 end
 
-"""
-Returns the change in density `δρ` for a given `δV`. Drop all non-diagonal terms with
-(f(εn)-f(εm))/(εn-εm) factor less than `droptol`. If `sternheimer_contribution`
-is false, only compute excitations inside the provided orbitals.
+# Returns δψ, δocc, δεF corresponding to a change in *total* Hamiltonian dH
+# P = f_fd(H-εF), tr(P) = N
+# δεn = <ψn|δV|ψn>
+# 0 = ∑_n fn' (δεn - δεF) determines δεF
+# where fn' = f_fd'((εn-εF)/T)/T
 
-Note: This function assumes that all bands contained in `ψ` and `eigenvalues` are
-sufficiently converged. By default the `self_consistent_field` routine of `DFTK`
-returns `3` extra bands, which are not converged by the eigensolver
-(see `n_ep_extra` parameter). These should be discarded before using this function.
-"""
-@views @timing function apply_χ0(ham, ψ, εF, eigenvalues, δV;
-                          droptol=0,
-                          sternheimer_contribution=true,
-                          temperature=ham.basis.model.temperature,
-                          kwargs_sternheimer=(cgtol=1e-6, verbose=false))
+# Then <ψm|δP|ψn> = (fm-fn)/(εm-εn) <ψm|δH|ψn>
+# Except for the diagonal which is
+# <ψn|δP|ψn> = (fn'-δεF) <ψn|δH|ψn>
+
+# We want to represent this with a δψ, δf, we do *not* impose that
+# δψ is orthogonal at finite temperature.
+# We get
+# δP = ∑_k fk (|δψk><ψk| + |ψk><δψk|) + δfk |ψk><ψk|
+# Identifying with <ψm|δP|ψn> we get for the diagonal terms
+# δfn = fn'<ψn|δH-δεF|ψn>
+# and for the off-diagonal
+# (fm-fn)/(εm-εn) <ψm|δH|ψn> = fm <δψm|ψn> + fn <ψm|δψn>
```

Comment:
I think I would add another paragraph break here (now you concern the explict terms with sternheimer and without)